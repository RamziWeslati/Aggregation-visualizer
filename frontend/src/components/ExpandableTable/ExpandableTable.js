import React, { useEffect, useState } from "react";
import PropTypes from "prop-types";

import {
  Table,
  TableBody,
  TableCell,
  TableContainer,
  TableHead,
  TableRow,
  LinearProgress,
  TablePagination,
} from "@material-ui/core";

import SimpleTable from "../Table/Table";
import Collapsable from "../Collapsable/Collapsable";

import { StyledTableCell, useStyles } from "../../styles/ExpandableTable.theme";

const ExpandableTable = ({ data, onExpand, header }) => {
  /**Component to render Table expandable rows of data
   * 
   * uses SimpleTable and Collapsable components 
   * 
   * @props
   * - data: array of object
   *      data to display, each object is mapped to a row
   * - header [optional]: array of string
   *      header to display as column names. defaulted to keys of a data object
   * - onExpand: function
   *      callback function to fetch expanded table data by
   * @state
   * - expandedRow / setExpandedRow: integer
   *      index of expanded row, null if table is not expanded
   * - expandedData / setExpandedData: array of object
   * - isLoading: boolean
   *      boolean to render loading components
   */

  // column names definition
  const columns = header ? header : Object.keys(data[0]);

  const [isLoading, setIsLoading] = useState(false);
  const [expandedRow, setExpandedRow] = useState(null);
  const [expandedData, setExpandedData] = useState([]);
  //pagination helpers
  const [page, setPage] = React.useState(0);
  const [rowsPerPage, setRowsPerPage] = React.useState(10);

  // fetch data to fill expanded child table, when we have a new expanded row
  useEffect( async () => {
    if (expandedRow !== null) {
      setIsLoading(true);
      const _expandedData = await onExpand(data[expandedRow])
      setExpandedData(_expandedData);
      setIsLoading(false);
    }
  }, [expandedRow]);

  const onCellClick = (idx) => {
    if (expandedRow == idx) {
      //undo expand and reset expanded data
      setExpandedRow(null);
      setExpandedData([]);
    }
    //expand
    else setExpandedRow(idx);
  };

  const handleChangePage = (event, newPage) => {
    // helper pagination function
    setPage(newPage);
  };

  const handleChangeRowsPerPage = (event) => {
    // helper pagination function
    setRowsPerPage(+event.target.value);
    setPage(0);
  };

  // render empty rows for the fill the fixed height
  const emptyRows =
    rowsPerPage -
    Math.min(
      rowsPerPage,
      data.length + expandedData.length - page * rowsPerPage
    ) // TODO compute the correct value to add

  const classes = useStyles()

  return (
    <>
      <TableContainer className={classes.container}>
        <Table stickyHeader aria-label="sticky table">
          <TableHead>
            <TableRow data-testid="header" >
              {columns.map((column, idx) => (
                <StyledTableCell key={`header-cell-${idx}`}>
                  {column}
                </StyledTableCell>
              ))}
            </TableRow>
          </TableHead>
          <TableBody>
            {(rowsPerPage > 0
              ? data.slice(page * rowsPerPage, page * rowsPerPage + rowsPerPage)
              : data
            ).flatMap((rowData, rIdx) => (
              <>
                <TableRow data-testid="row" key={`table-row-${rIdx}`}>
                  {Object.values(rowData).map((rowValue, cIdx) => (
                    <TableCell
                      key={`table-row-${cIdx}-${rIdx}`}
                      onClick={() => onCellClick(rIdx)}
                      className={classes.tableContentCell}
                    >
                      {rowValue}
                    </TableCell>
                  ))}
                </TableRow>
                {rIdx == expandedRow && ( isLoading? <LinearProgress  style={{width: '2000px'}} />
                :
                 expandedData.length > 0 && (
                  <Collapsable
                    data-testid="expanded-table"
                    colNumber={columns.length}
                  >
                    <SimpleTable data={expandedData} />
                  </Collapsable>
                ))}
              </>
            ))}
            {emptyRows > 0 && (
              <TableRow style={{ height: 53 * emptyRows }}>
                <TableCell colSpan={6} />
              </TableRow>
            )}
          </TableBody>
        </Table>
      </TableContainer>
      <TablePagination
        rowsPerPageOptions={[10, 25, 100]}
        component="div"
        count={data.length}
        rowsPerPage={rowsPerPage}
        page={page}
        onChangePage={handleChangePage}
        onChangeRowsPerPage={handleChangeRowsPerPage}
      />
    </>
  );
};

ExpandableTable.propTypes = {
  // custom prop checker to require non empty arrays of data
  data: (props, propName, componentName) => {
    const errorBaseString = `Invalid prop ${propName} supplied to ${componentName}, `;
    const dataProp = props[propName];

    if (dataProp === undefined)
      return new Error(errorBaseString + "data prop is required.");
    if (!Array.isArray(dataProp))
      return new Error(errorBaseString + "data prop must be an array.");
    if (dataProp.length == 0)
      return new Error(errorBaseString + "data prop must not be empty.");
  },
  onExpand: PropTypes.func.isRequired,
  header: PropTypes.array,
};

export default ExpandableTable;