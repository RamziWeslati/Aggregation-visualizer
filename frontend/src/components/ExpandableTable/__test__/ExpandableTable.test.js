import React from 'react';
import ReactDom from 'react-dom';

import { render, fireEvent } from '@testing-library/react';
import * as renderer from 'react-test-renderer';

import ExpandableTable from './../ExpandableTable';

const row1 = {'c11': 'v11', 'c12': 'v12', 'c13': 'v13'}
const row2 = {'c21': 'v21', 'c22': 'v22', 'c23': 'v23'}

const expandedRow1 = {'ec11': 'ev11', 'ec12': 'ev12', 'ec13': 'ev13'}
const expandedRow2 = {'ec21': 'ev21', 'ec22': 'ev22', 'ec23': 'ev23'}
const expandedRows = [expandedRow1, expandedRow2]

const mockExpand = () => jest.fn().mockResolvedValue({data: expandedRows})

const RequiredProps = {
    data: [row1, row2],
    onExpand: jest.fn(),
}

const optionalProps = {
    headers: ['Column Name 1', 'Column Name 2', 'Column Name 3']
}

const props = {...optionalProps, ...RequiredProps}

// cleanup is automated

// uts
// SimpleTable uts
it('renders without crashing, only required props', () => {
    const div = document.createElement('div')
    ReactDom.render(<ExpandableTable {...RequiredProps} />, div)
})

it('renders without crashing, full props', () => {
    const div = document.createElement('div')
    ReactDom.render(<ExpandableTable {...props} />, div)
})

it(`renders ${props.data.length} rows`, () => {
    const { getAllByTestId } = render(<ExpandableTable {...props} />)
    expect(getAllByTestId('row')).toHaveLength(props.data.length)
})

it(`renders ${props.data.length + 2} row role`, () => { // header and data rows
    const { getAllByRole } = render(<ExpandableTable {...props} />)
    expect(getAllByRole('row')).toHaveLength(props.data.length + 2)  // + 1 for header row
})

it(`renders a header when header is passed`, () => {
    const { getByTestId } = render(<ExpandableTable {...props} />)
    expect(getByTestId('header')).not.toBeNull()
})

it(`renders an infered header when header is not passed`, () => {
    const { getByTestId } = render(<ExpandableTable {...RequiredProps} />)
    expect(getByTestId('header')).not.toBeNull()
})

// Expand uts
it('calls onExpand when value cell is clicked', () => {
    const _props = {...props, onExpand: mockExpand()}
    const { getByText } = render(<ExpandableTable {..._props} />)
    const cell = getByText(Object.values(row1)[0])
    fireEvent.click(cell)
    expect(_props.onExpand).toBeCalledTimes(1)
})


// it('renders a nested Table within the table when row is clicked', () => {
//     const _props = {...props, onExpand: mockExpand()}
//     const { getAllByTestId, getByText } = render(<ExpandableTable {..._props} />)
//     const cell = getByText(Object.values(row1)[0])
//     fireEvent.click(cell)
//     expect(getAllByTestId('row')).toHaveLength(props.data.length + expandedRows.length)
// })


// snapshots
// un-expanded table snapshots

        //TODO, refactor this into removing UUIDs from snapshot comparaison

/*
 it('matches infered header snapshot, unextended', () => {
     const tree = renderer.create(<ExpandableTable {...RequiredProps} />).toJSON();
     expect(tree).toMatchSnapshot();
 })

it('matches custom header snapshot, unextended', () => {
    const tree = renderer.create(<ExpandableTable {...props} />).toJSON();
    expect(tree).toMatchSnapshot();
})

 expanded table snapshots

 it('matches nested table snapshot, extended', () => {
     const _props = {...props, onExpand: mockExpand()}
     const {getByText, asFragment} = render(<ExpandableTable {..._props} />);
     const cell = getByText(Object.values(row1)[0])
     fireEvent.click(cell)
     expect(asFragment()).toMatchSnapshot();
 })
 */