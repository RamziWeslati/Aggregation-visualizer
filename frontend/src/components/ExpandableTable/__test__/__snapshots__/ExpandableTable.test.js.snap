// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`matches custom header snapshot, unextended 1`] = `
Array [
  <div
    className="MuiTableContainer-root makeStyles-container-1"
  >
    <table
      aria-label="sticky table"
      className="MuiTable-root MuiTable-stickyHeader"
      role={null}
    >
      <thead
        className="MuiTableHead-root"
        role={null}
      >
        <tr
          className="MuiTableRow-root MuiTableRow-head"
          data-testid="header"
          role={null}
        >
          <th
            aria-sort={null}
            className="MuiTableCell-root MuiTableCell-head WithStyles(ForwardRef(TableCell))-head-3 MuiTableCell-stickyHeader"
            scope="col"
          >
            c11
          </th>
          <th
            aria-sort={null}
            className="MuiTableCell-root MuiTableCell-head WithStyles(ForwardRef(TableCell))-head-3 MuiTableCell-stickyHeader"
            scope="col"
          >
            c12
          </th>
          <th
            aria-sort={null}
            className="MuiTableCell-root MuiTableCell-head WithStyles(ForwardRef(TableCell))-head-3 MuiTableCell-stickyHeader"
            scope="col"
          >
            c13
          </th>
        </tr>
      </thead>
      <tbody
        className="MuiTableBody-root"
        role={null}
      >
        <tr
          className="MuiTableRow-root"
          data-testid="row"
          role={null}
        >
          <td
            aria-sort={null}
            className="MuiTableCell-root MuiTableCell-body makeStyles-tableContentCell-2"
            onClick={[Function]}
          >
            v11
          </td>
          <td
            aria-sort={null}
            className="MuiTableCell-root MuiTableCell-body makeStyles-tableContentCell-2"
            onClick={[Function]}
          >
            v12
          </td>
          <td
            aria-sort={null}
            className="MuiTableCell-root MuiTableCell-body makeStyles-tableContentCell-2"
            onClick={[Function]}
          >
            v13
          </td>
        </tr>
        <tr
          className="MuiTableRow-root"
          data-testid="row"
          role={null}
        >
          <td
            aria-sort={null}
            className="MuiTableCell-root MuiTableCell-body makeStyles-tableContentCell-2"
            onClick={[Function]}
          >
            v21
          </td>
          <td
            aria-sort={null}
            className="MuiTableCell-root MuiTableCell-body makeStyles-tableContentCell-2"
            onClick={[Function]}
          >
            v22
          </td>
          <td
            aria-sort={null}
            className="MuiTableCell-root MuiTableCell-body makeStyles-tableContentCell-2"
            onClick={[Function]}
          >
            v23
          </td>
        </tr>
        <tr
          className="MuiTableRow-root"
          role={null}
          style={
            Object {
              "height": 424,
            }
          }
        >
          <td
            aria-sort={null}
            className="MuiTableCell-root MuiTableCell-body"
            colSpan={6}
          />
        </tr>
      </tbody>
    </table>
  </div>,
  <div
    className="MuiTablePagination-root"
  >
    <div
      className="MuiToolbar-root MuiToolbar-regular MuiTablePagination-toolbar MuiToolbar-gutters"
    >
      <div
        className="MuiTablePagination-spacer"
      />
      <p
        className="MuiTypography-root MuiTablePagination-caption MuiTypography-body2 MuiTypography-colorInherit"
        id="mui-22081"
      >
        Rows per page:
      </p>
      <div
        className="MuiInputBase-root MuiTablePagination-input MuiTablePagination-selectRoot"
        onClick={[Function]}
      >
        <div
          aria-haspopup="listbox"
          aria-labelledby="mui-22081 mui-53843"
          className="MuiSelect-root MuiSelect-select MuiTablePagination-select MuiSelect-selectMenu MuiInputBase-input"
          id="mui-53843"
          onBlur={[Function]}
          onFocus={[Function]}
          onKeyDown={[Function]}
          onMouseDown={[Function]}
          role="button"
          tabIndex={0}
        >
          10
        </div>
        <input
          aria-hidden={true}
          className="MuiSelect-nativeInput"
          onAnimationStart={[Function]}
          onChange={[Function]}
          tabIndex={-1}
          value={10}
        />
        <svg
          aria-hidden={true}
          className="MuiSvgIcon-root MuiSelect-icon MuiTablePagination-selectIcon"
          focusable="false"
          viewBox="0 0 24 24"
        >
          <path
            d="M7 10l5 5 5-5z"
          />
        </svg>
      </div>
      <p
        className="MuiTypography-root MuiTablePagination-caption MuiTypography-body2 MuiTypography-colorInherit"
      >
        1-2 of 2
      </p>
      <div
        className="MuiTablePagination-actions"
      >
        <button
          aria-label="Previous page"
          className="MuiButtonBase-root MuiIconButton-root MuiIconButton-colorInherit Mui-disabled Mui-disabled"
          disabled={true}
          onBlur={[Function]}
          onClick={[Function]}
          onDragLeave={[Function]}
          onFocus={[Function]}
          onKeyDown={[Function]}
          onKeyUp={[Function]}
          onMouseDown={[Function]}
          onMouseLeave={[Function]}
          onMouseUp={[Function]}
          onTouchEnd={[Function]}
          onTouchMove={[Function]}
          onTouchStart={[Function]}
          tabIndex={-1}
          title="Previous page"
          type="button"
        >
          <span
            className="MuiIconButton-label"
          >
            <svg
              aria-hidden={true}
              className="MuiSvgIcon-root"
              focusable="false"
              viewBox="0 0 24 24"
            >
              <path
                d="M15.41 16.09l-4.58-4.59 4.58-4.59L14 5.5l-6 6 6 6z"
              />
            </svg>
          </span>
        </button>
        <button
          aria-label="Next page"
          className="MuiButtonBase-root MuiIconButton-root MuiIconButton-colorInherit Mui-disabled Mui-disabled"
          disabled={true}
          onBlur={[Function]}
          onClick={[Function]}
          onDragLeave={[Function]}
          onFocus={[Function]}
          onKeyDown={[Function]}
          onKeyUp={[Function]}
          onMouseDown={[Function]}
          onMouseLeave={[Function]}
          onMouseUp={[Function]}
          onTouchEnd={[Function]}
          onTouchMove={[Function]}
          onTouchStart={[Function]}
          tabIndex={-1}
          title="Next page"
          type="button"
        >
          <span
            className="MuiIconButton-label"
          >
            <svg
              aria-hidden={true}
              className="MuiSvgIcon-root"
              focusable="false"
              viewBox="0 0 24 24"
            >
              <path
                d="M8.59 16.34l4.58-4.59-4.58-4.59L10 5.75l6 6-6 6z"
              />
            </svg>
          </span>
        </button>
      </div>
    </div>
  </div>,
]
`;

exports[`matches infered header snapshot, unextended 1`] = `
Array [
  <div
    className="MuiTableContainer-root makeStyles-container-1"
  >
    <table
      aria-label="sticky table"
      className="MuiTable-root MuiTable-stickyHeader"
      role={null}
    >
      <thead
        className="MuiTableHead-root"
        role={null}
      >
        <tr
          className="MuiTableRow-root MuiTableRow-head"
          data-testid="header"
          role={null}
        >
          <th
            aria-sort={null}
            className="MuiTableCell-root MuiTableCell-head WithStyles(ForwardRef(TableCell))-head-3 MuiTableCell-stickyHeader"
            scope="col"
          >
            c11
          </th>
          <th
            aria-sort={null}
            className="MuiTableCell-root MuiTableCell-head WithStyles(ForwardRef(TableCell))-head-3 MuiTableCell-stickyHeader"
            scope="col"
          >
            c12
          </th>
          <th
            aria-sort={null}
            className="MuiTableCell-root MuiTableCell-head WithStyles(ForwardRef(TableCell))-head-3 MuiTableCell-stickyHeader"
            scope="col"
          >
            c13
          </th>
        </tr>
      </thead>
      <tbody
        className="MuiTableBody-root"
        role={null}
      >
        <tr
          className="MuiTableRow-root"
          data-testid="row"
          role={null}
        >
          <td
            aria-sort={null}
            className="MuiTableCell-root MuiTableCell-body makeStyles-tableContentCell-2"
            onClick={[Function]}
          >
            v11
          </td>
          <td
            aria-sort={null}
            className="MuiTableCell-root MuiTableCell-body makeStyles-tableContentCell-2"
            onClick={[Function]}
          >
            v12
          </td>
          <td
            aria-sort={null}
            className="MuiTableCell-root MuiTableCell-body makeStyles-tableContentCell-2"
            onClick={[Function]}
          >
            v13
          </td>
        </tr>
        <tr
          className="MuiTableRow-root"
          data-testid="row"
          role={null}
        >
          <td
            aria-sort={null}
            className="MuiTableCell-root MuiTableCell-body makeStyles-tableContentCell-2"
            onClick={[Function]}
          >
            v21
          </td>
          <td
            aria-sort={null}
            className="MuiTableCell-root MuiTableCell-body makeStyles-tableContentCell-2"
            onClick={[Function]}
          >
            v22
          </td>
          <td
            aria-sort={null}
            className="MuiTableCell-root MuiTableCell-body makeStyles-tableContentCell-2"
            onClick={[Function]}
          >
            v23
          </td>
        </tr>
        <tr
          className="MuiTableRow-root"
          role={null}
          style={
            Object {
              "height": 424,
            }
          }
        >
          <td
            aria-sort={null}
            className="MuiTableCell-root MuiTableCell-body"
            colSpan={6}
          />
        </tr>
      </tbody>
    </table>
  </div>,
  <div
    className="MuiTablePagination-root"
  >
    <div
      className="MuiToolbar-root MuiToolbar-regular MuiTablePagination-toolbar MuiToolbar-gutters"
    >
      <div
        className="MuiTablePagination-spacer"
      />
      <p
        className="MuiTypography-root MuiTablePagination-caption MuiTypography-body2 MuiTypography-colorInherit"
      >
        Rows per page:
      </p>
      <div
        className="MuiInputBase-root MuiTablePagination-input MuiTablePagination-selectRoot"
        onClick={[Function]}
      >
        <div
          aria-haspopup="listbox"
          className="MuiSelect-root MuiSelect-select MuiTablePagination-select MuiSelect-selectMenu MuiInputBase-input"
          onBlur={[Function]}
          onFocus={[Function]}
          onKeyDown={[Function]}
          onMouseDown={[Function]}
          role="button"
          tabIndex={0}
        >
          10
        </div>
        <input
          aria-hidden={true}
          className="MuiSelect-nativeInput"
          onAnimationStart={[Function]}
          onChange={[Function]}
          tabIndex={-1}
          value={10}
        />
        <svg
          aria-hidden={true}
          className="MuiSvgIcon-root MuiSelect-icon MuiTablePagination-selectIcon"
          focusable="false"
          viewBox="0 0 24 24"
        >
          <path
            d="M7 10l5 5 5-5z"
          />
        </svg>
      </div>
      <p
        className="MuiTypography-root MuiTablePagination-caption MuiTypography-body2 MuiTypography-colorInherit"
      >
        1-2 of 2
      </p>
      <div
        className="MuiTablePagination-actions"
      >
        <button
          aria-label="Previous page"
          className="MuiButtonBase-root MuiIconButton-root MuiIconButton-colorInherit Mui-disabled Mui-disabled"
          disabled={true}
          onBlur={[Function]}
          onClick={[Function]}
          onDragLeave={[Function]}
          onFocus={[Function]}
          onKeyDown={[Function]}
          onKeyUp={[Function]}
          onMouseDown={[Function]}
          onMouseLeave={[Function]}
          onMouseUp={[Function]}
          onTouchEnd={[Function]}
          onTouchMove={[Function]}
          onTouchStart={[Function]}
          tabIndex={-1}
          title="Previous page"
          type="button"
        >
          <span
            className="MuiIconButton-label"
          >
            <svg
              aria-hidden={true}
              className="MuiSvgIcon-root"
              focusable="false"
              viewBox="0 0 24 24"
            >
              <path
                d="M15.41 16.09l-4.58-4.59 4.58-4.59L14 5.5l-6 6 6 6z"
              />
            </svg>
          </span>
        </button>
        <button
          aria-label="Next page"
          className="MuiButtonBase-root MuiIconButton-root MuiIconButton-colorInherit Mui-disabled Mui-disabled"
          disabled={true}
          onBlur={[Function]}
          onClick={[Function]}
          onDragLeave={[Function]}
          onFocus={[Function]}
          onKeyDown={[Function]}
          onKeyUp={[Function]}
          onMouseDown={[Function]}
          onMouseLeave={[Function]}
          onMouseUp={[Function]}
          onTouchEnd={[Function]}
          onTouchMove={[Function]}
          onTouchStart={[Function]}
          tabIndex={-1}
          title="Next page"
          type="button"
        >
          <span
            className="MuiIconButton-label"
          >
            <svg
              aria-hidden={true}
              className="MuiSvgIcon-root"
              focusable="false"
              viewBox="0 0 24 24"
            >
              <path
                d="M8.59 16.34l4.58-4.59-4.58-4.59L10 5.75l6 6-6 6z"
              />
            </svg>
          </span>
        </button>
      </div>
    </div>
  </div>,
]
`;
